cmake/templates/ros_bin.cpp

```diff
@@ -0,0 +1,110 @@
+/**
+ * @file /src/ros_bin.cpp
+ *
+ * @brief Self describing executable for launching python scripts in /bin
+ *
+ * @date March 2011

```

Comment:
It seems this was copied from somewhere. In that case we need to check and state the copyright and license.

Reply:
Depending on this I will probably add further comments (e.g. code style, some renaming, error handling).

Reply:
this file was copied and adapted from [stonier/win_ros](https://github.com/stonier/win_ros/blob/fuerte/msvc/msvc_runtime/src/ros_bin.cpp), @stonier could you please help us check how we could/should add copyright/license info for this file?

Reply:
I'm adding a BSD License template first as a placeholder

Reply:
Waiting for an explicit confirmation of the original author @stonier before moving forward with the review.

Reply:
That's a blast from the past. Wondering where you dug up that header from? The original [file](https://github.com/stonier/win_ros/blob/master/msvc/msvc_runtime/src/ros_bin.cpp) doesn't have any license, nor does the repository itself, so the TLDR; use it in whatever way you wish, there are no legal copyright references there.

Much of that code was written back before we/anyone at Yujin paid attention to licenses (first repository they permitted to be freely available, and there was no intent to have that one locked down at all).

Reply:
Thanks for jumping in @stonier! I think Stuart from our team was the one who found this, but yeah, good code only gets old not lost =)

I'm pretty new to all the copyright knowledge, so I'll just throw these questions out:
- does this mean we would not need any kind of license for this file?
- (if no license is added,) will this be potentially problematic in the future?
- do we want to use a BSD license if we indeed need a license? BSD seems to be pretty open, we might want to remove the 3rd clause in the one I added though.

Reply:
The original repository is neither properly licensed nor copyrighted. Do as you wish with it.

Having said that, building wrappers for each and every python script doesn't seem like the right way forward here - it doesn't scale and won't be under continuous integration. Wrappers were used for `win_ros` since that was both experimental and targeting a very bounded set of packages sufficient for connecting to and interacting with a full linux-based ROS robot. Getting ROS1 to support windows natively, across the board was beyond the resources we were willing to apply without OSRF and the build farm behind it.

If there is another way, that surely has to be preferable. Setuptools packages? Associations (a'la  [python3 shebang support](https://docs.python.org/3/using/windows.html#shebang-lines)? You might even resort to compiling wrappers separately - either on the fly by scanning an installed system, or building a package of wrappers that can be installed, but raise exceptions if the wrapped script itself isn't installed.

Whatever the case, life should be a bit easier now than when we struggled with WinXP.

Reply:
good to know! I think having a license in the wrapper `.cpp` file would make like easier for OSRF to maintain it later on in the future (instead of keeping a potential copyright argument). I would suggest keep the line of `Code courtesy of Yujin Robot.`, yet in this case, maybe update the license to have either `Willow Garage, Inc.` or `OSRF` in it.

and thanks a lot for sharing your valuable knowledge! just to answer some of the mentioned questions:

> building wrappers for each and every python script doesn't seem like the right way forward here

totally agree! manually adding these
```
foreach(script ${PYTHON_SCRIPTS})
  add_python_executable_helper(
    SCRIPT_NAME ${script}
    TARGET_NAME ${script}_executable)
endforeach()
```
to each package would be super painful and practically unfeasible. I'm in the middle of another commit to move these into `catkin` wrappers so whenever a package wants to deploy a script (`scripts=[...]` in `setup.py`) or install a script as its node (`catkin_install_python`), `add_python_executable_helper` would be called.

this, however, is the best we could do at this point. there could be further improvement, but that would have to come in the future =)

> Getting ROS1 to support windows natively, across the board was beyond the resources we were willing to apply without OSRF and the build farm behind it.

working on this with OSRF! while it might be hard for ROS1 to get fully supported on Windows, good news is that ROS2 is going to be (in design)

> `setuptools` package?

I did a bit of investigation and prototyping (while my bandwidth would allow) on my own, and wrote down findings as a comment in this pull request. Ideally `entry_point::console_script` should work; however, during the porting of ROS to Windows, we encountered the issue of user already has existing Python installation. To work around that, our current approach is to add a customized Python installation in `C:\opt\python27amd64` (fully hard-coded path). `entry_point::console_script` from `setuptools` does create a win32 wrapper on Windows, yet it tries to follow the shebang line support on Linux and will always try to execute the script with the specific Python installation when the executable is generated (Python from `C:\opt\python27amd64`). There is no such guarantee that Python (for ROS) will always be installed to this path, or if the user would be using Python like that. This approach could make maintaining and upgrading a huge pain.

> Associations (a'la python3 shebang support)?

this would require a system-level change, and we are trying to avoid any change on that level. very easy to get affected by environment corruption

> compiling wrappers separately

this seems to be what we (also the original `win_ros` project) are doing. the wrapper itself does nothing but try to execute a Python script with the registered Python interpreter. In my opinion, this is not done on the fly for an ROS installation because this would mean the user needs to do an extra step before using the ROS installation

our vision is to make ROS installation as stream-line as possible, so it could appear as an enhancement =)

> life should be a bit easier now than when we struggled with WinXP

Indeed! the popular `win_ros` project has been a great source of knowledge with this wrapper solution and a lot more, and ROS itself is gradually becoming more and more Windows-friendly. A lot of changes in ROS2 have been great examples of how ROS1 should be ported to Windows =)